### PR TITLE
Add user interests to settings & profile page

### PR DESCRIPTION
**Pull request by meredithcat**
## Description

**Issue:** #627 

One potential issue is that I had to add the $uiLanguage variable to a bunch of other user queries, so that the user interests could be added to the queries. If this is not desired, then we could create a new user fragment that contains the interests, specifically for the pages where it is needed, but that seemed like overkill.

Still to do is adding a custom filter for "My Interests" to the feed page, so let me know if I should include that in this PR or a future one.

## Subtasks

- [x] I have added this PR to the Journaly Kanban project ✅

### Deployment Checklist

- [ ] 🚨 Publish new j-db-client version and update in both `web` and `j-mail`
- [ ] 🚨 Deploy migs to stage
- [ ] 🚨 Deploy code to stage
- [ ] 🚨 DEPLOY `j-mail` to stage
- [ ] 🚨 Deploy migs to prod
- [ ] 🚨 Deploy code to prod
- [ ] 🚨 DEPLOY `j-mail` TO PROD

## Screenshots

Screen Shot 2021-08-03 at 7 31 08 PM
Screen Shot 2021-08-03 at 7 45 37 PM

### CODE REVIEW

**Comment by robin-macpherson** [comment]
Woohoo can't wait to take a closer look today 🎉 Regarding the TODO, let's not do that until the work on #617 is finished which should be in the next couple of days

**Comment by Lanny** [comment]
> If this is not desired, then we could create a new user fragment that contains the interests, specifically for the pages where it is needed

I think this would be my preferred approach, just because it reduces the amount of bureaucracy most other pages have to go through. 

Also I’d suggest making the mutations return a user type instead of a topic type (and move them to the user file accordingly), that way there wouldn’t be a need to call ‘refetch()’ after the mutation, since the mutation itself will return an updated user record and Apollo should be able to reconcile the cache on its own, saving a roundtrip to the server. 

That optimization aside, the approach looks good 👍 , good stuff!

**Review by robin-macpherson — COMMENTED** [review]
Amazing work @meredithcat ! I have the same comments as @Lanny - typically we favor the approach of creating a fragment in this scenario which you'll see a lot of around the codebase as you explore it more. Happy to approve and get this merged in once those optimizations are made, which is really exciting 🥳

**Comment by meredithcat** [comment]
Done - looks like I actually didn't need an extra fragment here - now the `UserInterestFragment` is only being used in the profile & settings pages.

**Comment by meredithteo** [comment]
@robin-macpherson @Lanny Do you all have bandwidth to finish up reviewing this change? If so, I think it is close to being ready for merge. 🚀

**Comment by robin-macpherson** [comment]
@meredithcat I'm drowning in stuff at the moment but we'll definitely find the time to get your two PRs through this coming week. Thanks for taking care of the requested changes here.

File: packages/web/resolvers/topic.ts [hunk 1]
```diff
@@ -119,4 +128,68 @@ const TopicQueries = extendType({
...
+        const interestFilter = {
+          where: {
+            userId_topicId: {
+              topicId: args.topicId,
+              userId,
+            },
+          },
+        }
+
+        const userInterest = await ctx.db.userInterest.findUnique(interestFilter)
```
Comment (Lanny) [inline packages/web/resolvers/topic.ts:183]: Is checking for the existence of the user/topic relation. necessary? We fail if it doesn't exist, why not just attempt the delete and fail if the delete doesn't go through? That way we could save a round trip to the DB

**Review by Lanny — COMMENTED** [review]
Looks good, added one comment where I think we can save a DB query but otherwise seems good to me.

File: packages/web/resolvers/topic.ts (hunk 1 above)
Comment (meredithteo) [reply packages/web/resolvers/topic.ts:183]: hmm, this is also done here: https://github.com/Journaly/journaly/blob/master/packages/web/resolvers/language.ts#L139

let me know if you would still advise to remove this check!

Comment (Lanny) [reply packages/web/resolvers/topic.ts:183]: Yeah, in both cases I think we should just do the delete. `db.delete()` will throw if no record is found for deletion, so we can either just let that exception bubble up or catch and rethrow our own exception, it's the same result either way.

**Comment by Lanny** [comment]
Type checker is complaining (not sure if you can see the circle build logs, but `npm run type-check` or `npm run build` should produce the same message):

```
./pages/dashboard/settings/profile.tsx:43:21
Type error: Type '({ __typename?: "UserInterest" | undefined; } & { __typename?: "UserInterest" | undefined; } & { topic: { __typename?: "Topic" | undefined; } & { __typename?: "Topic" | undefined; } & Pick<...>; })[] | undefined' is not assignable to type 'UserInterest[]'.
  Type 'undefined' is not assignable to type 'UserInterest[]'.

  41 |                   <InterestsForm
  42 |                     topics={data.topics}
> 43 |                     userInterests={data.currentUser?.userInterests}
     |                     ^
  44 |                     refetch={refetch}
  45 |                   />
  46 |                   <SocialForm socialMedia={data.currentUser?.socialMedia} refetch={refetch} />
```

This is kind of an annoying situation (which is not your fault). We're checking that the `currentUser` object returned by `AuthGate` is not undefined, but not checking that `data.currentUser` is not undefined. I don't know that they could ever realistically have different values (maybe if the relevant queries load separately) but TSC can't prove it if so. I think the immediate solution is to check `data.currentUser` along side checking `currentUser` here: https://github.com/Journaly/journaly/pull/633/files#diff-f8b3bc26037afc850349067e596a1f2f4d21e340f5c6f141a5a6d499bbe0b431R29

**Comment by robin-macpherson** [comment]
@Lanny I went ahead and addressed your requested changes re: removing those relation checks and also fixed the build. While testing I also found an existing bug where the selected ID state was not being cleared after adding a language or interest, which resulted in the submit button remaining active and the user could click it again, causing a Unique Constraint error thrown on the server (video below). Each fix in separate commits from today, if you can take a look and let me know if all looks good to you that would be awesome 🙏🏼

https://user-images.githubusercontent.com/34203886/135686673-fca641d9-d318-4cb3-aab0-c31475b10a5d.mov

**Review by Lanny — APPROVED** [review]
👍 good stuff

**Pull request merged**